csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/common/TranslatedCallBlueprint.qll

```diff
@@ -0,0 +1,217 @@
+    (
+      tag = CallTag() and
+      (
+        (
+          operandTag instanceof CallTargetOperandTag and
+          result = getInstruction(CallTargetTag())
+        ) or
+        (
+          operandTag instanceof ThisArgumentOperandTag and
+          result = getQualifierResult()
+        ) or
+        exists(PositionalArgumentOperandTag argTag |
+          argTag = operandTag and
+          result = getArgument(argTag.getArgIndex()).getResult()
+        )
+      )
+    ) or
+    (
+      tag = CallSideEffectTag() and
+      hasSideEffect() and
+      operandTag instanceof SideEffectOperandTag and
+      result = getUnmodeledDefinitionInstruction()
+    )
+  }
+
+  override final Type getInstructionOperandType(InstructionTag tag,
+      TypedOperandTag operandTag) {
+    tag = CallSideEffectTag() and
+    hasSideEffect() and
+    operandTag instanceof SideEffectOperandTag and
+    result instanceof Language::UnknownType
+  }
+
+  Instruction getResult() {
+    result = getInstruction(CallTag())
+  }
+
+  /**
+   * Gets the result type of the call.
+   */
+  abstract Type getCallResultType();
+
+  /**
+   * Gets the unmodeled definition instruction of the enclosing
+   * function (of the element this call is attached to).
+   */
+  abstract Instruction getUnmodeledDefinitionInstruction();
+  
+  /**
+   * Holds if the call has a `this` argument.
+   */
+  predicate hasQualifier() {
+    exists(getQualifier())
+  }
+
+  /**
+   * Gets the expr for the qualifier of the call. 
+   */
+  abstract TranslatedExprBlueprint getQualifier();
+
+  /**
+   * Gets the instruction whose result value is the `this` argument of the call.
+   * In general, this is just the result of `getQualifier()`, but it can be
+   * overridden by a subclass for cases where there is a `this` argument that is
+   * not computed from a child expression (e.g. a constructor call).
+   */
+  abstract Instruction getQualifierResult();
+
+  /**
+   * Gets the argument with the specified `index`. Does not include the `this`
+   * argument. We use `TranslatedExprBlueprint` so that we can give both `TranslatedExpr` args,
+   * in the case of AST generated arguments, or `TranslatedCompilerElement` args in the case of
+   * compiler generated arguments.
+   */
+  abstract TranslatedExprBlueprint getArgument(int index);
+
+  /**
+   * If there are any arguments, gets the first instruction of the first
+   * argument. Otherwise, returns the call instruction.
+   */
+  final Instruction getFirstArgumentOrCallInstruction() {
+    if hasArguments() then
+      result = getArgument(0).getFirstInstruction()
+    else
+      result = getInstruction(CallTag())
+  }
+
+  /**
+   * Holds if the call has any arguments, not counting the `this` argument.
+   */
+  abstract predicate hasArguments();
+
+  predicate hasReadSideEffect() {
+    any()
+  }
+
+  predicate hasWriteSideEffect() {
+    any()
+  }
+
+  private predicate hasSideEffect() {
+    hasReadSideEffect() or hasWriteSideEffect()
```

Comment:
What are examples of read/write side effects?

Reply:
Side effects are not really functional right now (they are WIP in the C++ IR).

Reply:
OK, still, what would be concrete examples?

Reply:
An example would be `out` or `ref` parameters passed as arguments to a function, since the values modified by the callee are reflected in the caller. Here is the PR that tackles side effects in the C++ IR: https://github.com/Semmle/ql/pull/1735. I haven't given much thought to how we'd tackle those in C#, I mainly left them as placeholders.